### PR TITLE
KT-45777: Take snapshots and compute changes for Kotlin classes

### PR DESCRIPTION
**Pull request by hungvietnguyen**

### CODE REVIEW

File: libraries/tools/kotlin-gradle-plugin/src/testData/kotlin.incremental.useClasspathSnapshot/changedMethodImplementation/com/example/A.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.example
+
+class A {
```
Comment (gavra0) [inline libraries/tools/kotlin-gradle-plugin/src/testData/kotlin.incremental.useClasspathSnapshot/changedMethodImplementation/com/example/A.kt:3]: Can you please add more test cases:
- private/non-private fields
- inline functions
- changing supertype by adding somethings that changes/does not change the supertype ABI
- adding an annotation

we do not need to test every edge case (there are a lot of incremental tests), but this is just to leverage unit tests as much as possible.

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathChangesComputer.kt [hunk 2]
```diff
@@ -65,20 +68,37 @@ object ClasspathChangesComputer {
...
-    private fun collectClassChanges(
-        current: ClassSnapshot,
-        previous: ClassSnapshot,
-        changesCollector: ChangesCollector
-    ): ChangesCollectorResult {
-        return Failure.NotYetImplemented
+    @TestOnly
+    internal fun collectKotlinClassChanges(current: KotlinClassSnapshot, previous: KotlinClassSnapshot): DirtyData {
+        val workingDir = FileUtil.createTempDirectory(this::class.java.simpleName, "WorkingDir", /* deleteOnExit */ true)
+        val incrementalJvmCache = IncrementalJvmCache(workingDir, /* targetOutputDir */ null, FileToCanonicalPathConverter)
```
Comment (gavra0) [inline libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathChangesComputer.kt:83]: Can IncrementalJvmCache store all classes from `ClasspathSnapshot`, and then we can create one in `getChanges` and pass it around to load/store class info? Right now, we create an instance per class. It is mostly lazy, but there is some overhead.

File: libraries/tools/kotlin-gradle-plugin/src/testData/kotlin.incremental.useClasspathSnapshot/changedMethodImplementation/com/example/A.kt (hunk 1 above)
Comment (hungvietnguyen) [reply libraries/tools/kotlin-gradle-plugin/src/testData/kotlin.incremental.useClasspathSnapshot/changedMethodImplementation/com/example/A.kt:3]: Yup, I've added a TODO here and will come back to it once we have a working version.

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathChangesComputer.kt (hunk 2 above)
Comment (hungvietnguyen) [reply libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathChangesComputer.kt:83]: Yes I planned to do that too, added a TODO. Right now my plan is:
 - Snapshot and diff correctly for a Kotlin class (this PR)
 - Snapshot and diff correctly for a Java class (next PR)
 - Connect everything up to get a working version and update integration tests, add unit tests (PR after that)

**Comment by hungvietnguyen** [comment]
Updated PR to simplify test data.

File: libraries/tools/kotlin-gradle-plugin/src/test/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathSnapshotTestCommon.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import org.junit.rules.TemporaryFolder
+import java.io.File
+
+abstract class ClasspathSnapshotTestCommon {
+
+    companion object {
+
+        val testDataDir = File("libraries/tools/kotlin-gradle-plugin/src/testData/kotlin.incremental.useClasspathSnapshot/src")
+
+        data class SourceFile(val baseDir: File, val relativePath: String) {
```
Comment (gavra0) [inline libraries/tools/kotlin-gradle-plugin/src/test/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathSnapshotTestCommon.kt:20]: nit: Pull out of the companion?

**Review by gavra0 — APPROVED** [review]

**Comment by yaya8754** [comment]
Thanks

**Comment by yaya8754** [comment]
30
101
140

File: libraries/tools/kotlin-gradle-plugin/src/test/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathSnapshotTestCommon.kt (hunk 3 above)
Comment (hungvietnguyen) [reply libraries/tools/kotlin-gradle-plugin/src/test/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathSnapshotTestCommon.kt]: Done

**Comment by hungvietnguyen** [comment]
Rebased PR to address review (no important changes). All tests should pass.

File: libraries/tools/kotlin-gradle-plugin/src/test/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathSnapshotTestCommon.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+        val newSourceFile = SourceFile(tmp.newFolder(), relativePath).also { it.asFile().parentFile.mkdirs() }
+        newSourceFile.asFile().writeText(fileContents.replace(oldValue, newValue))
+        return newSourceFile
+    }
+
+    protected fun SourceFile.compile(): ClassFile {
+        val classesDir = tmp.newFolder()
+        when {
+            relativePath.endsWith(".kt") -> compileKotlin(asFile().path, classesDir)
```
Comment (hungvietnguyen) [inline libraries/tools/kotlin-gradle-plugin/src/test/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathSnapshotTestCommon.kt:51]: @nav-nav @gavra0 The current PR's tests are failing at this code path. The reason is that this code uses the existing `org.jetbrains.kotlin.test.MockLibraryUtil#compileKotlin()` method to compile a Kotlin class. However, that method at runtime will require `<kotlin-repo>/dist/kotlinc/lib/kotlin-compiler.jar` to be present, and that requires running `./gradlew dist` first before running the tests.

I also tried to use `K2JVMCompiler` directly with the following code:
```
    private fun compileKotlin(sourceFile: File, classesDir: File) {
        val args = listOf(sourceFile.path, "-d", classesDir.path)
        val err = ByteArrayOutputStream()
        K2JVMCompiler().exec(PrintStream(err), *args.toTypedArray())
        err.toString().let {
            if (it.isNotEmpty()) {
                error(it)
            }
        }
    }
```
but it fails with `NoClassDefFoundError: com/intellij/core/JavaCoreApplicationEnvironment`

If you have ideas how to fix the above, please let me know. Otherwise, I plan to add the .class files to the test data for now.

Comment (hungvietnguyen) [reply libraries/tools/kotlin-gradle-plugin/src/test/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathSnapshotTestCommon.kt]: I have updated the PR with .class files in test data. This PR has now passed TeamCity tests. Nataliya, could you help trigger an RRA run? Thanks! @nav-nav

**Comment by hungvietnguyen** [comment]
Rebased to fix a merge conflict (trivial).

**Comment by hungvietnguyen** [comment]
This PR has been merged: https://github.com/JetBrains/kotlin/commit/206457d9ff452264562c0c3a30ef5784bad1ae5a, https://github.com/JetBrains/kotlin/commit/16dfdb620eeff7b6e1af1de62c63beb72a50e9fa
